packages/blockfrost/src/blockfrostProvider.ts

```diff
@@ -318,12 +327,92 @@ export const blockfrostProvider = (options: Options, logger = dummyLogger): Wall
     return BlockfrostToCore.currentWalletProtocolParameters(response.data);
   };
 
+  const accountRewards = async (
+    stakeAddress: Cardano.Address,
+    { lowerBound = 0, upperBound = Number.MAX_SAFE_INTEGER }: EpochRange = {}
+  ): Promise<EpochRewards[]> => {
+    const result: EpochRewards[] = [];
+    const batchSize = 100;
+    let page = 1;
+    let haveMorePages = true;
+    while (haveMorePages) {
+      const rewards = await blockfrost.accountsRewards(stakeAddress, { count: batchSize, page });
+      result.push(
+        ...rewards
+          .filter(({ epoch }) => lowerBound <= epoch && epoch <= upperBound)
+          .map(({ epoch, amount }) => ({
+            epoch,
+            rewards: BigInt(amount)
+          }))
+      );
+      haveMorePages = rewards[rewards.length - 1].epoch < upperBound && rewards.length === 100;
+      page += 1;
+    }
+    return result;
+  };
+
+  const rewardsHistory: WalletProvider['rewardsHistory'] = async ({ stakeAddresses, epochs }) => {
+    const allAddressRewards = await Promise.all(stakeAddresses.map((address) => accountRewards(address, epochs)));
+    const accountRewardsByEpoch = groupBy(flatten(allAddressRewards), ({ epoch }) => epoch);
+    return Object.keys(accountRewardsByEpoch).map((key) => ({
+      epoch: accountRewardsByEpoch[key][0].epoch,
+      rewards: BigIntMath.sum(accountRewardsByEpoch[key].map(({ rewards }) => rewards))
+    }));
+  };
+
+  const genesisParameters: WalletProvider['genesisParameters'] = async () => {
+    const response = await blockfrost.genesis();
+    return {
+      activeSlotsCoefficient: response.active_slots_coefficient,
+      epochLength: response.epoch_length,
+      maxKesEvolutions: response.max_kes_evolutions,
+      maxLovelaceSupply: BigInt(response.max_lovelace_supply),
+      networkMagic: response.network_magic,
+      securityParameter: response.security_param,
+      slotLength: response.slot_length,
+      slotsPerKesPeriod: response.slots_per_kes_period,
+      systemStart: new Date(response.system_start * 1000),
+      updateQuorum: response.update_quorum
+    };
+  };
+
+  const queryBlocksByHashes: WalletProvider['queryBlocksByHashes'] = async (hashes) => {
+    const responses = await Promise.all(hashes.map((hash) => blockfrost.blocks(hash)));
+    // Review: a bunch of nullable fields here. Is there a reason for them to be nullable?
+    // Maybe just throw if any of the fields are not set?
+    return responses.map((response) => ({
+      confirmations: response.confirmations,
+      date: new Date(response.time * 1000),
+      epoch: response.epoch!,
```

Comment:
There are two type of blocks that curiously do not belong to epochs. We do not need to be concerned with these special blocks though, as they are only applicable to the Byron-era and do not contain transactions. Can throw if not set